### PR TITLE
Respect errors from UserInfo

### LINKED ISSUE #9768: Respect 401 from UserInfo
When querying the OIDC UserInfo endpoint, Teleport only detects 404 at the moment.

https://github.com/gravitational/teleport/blob/95d0f0d27f2d7982d481e19a65adf914b6dca3f7/lib/auth/oidc.go#L795-L804

It appears that the OIDC spec supports 401 as well:

https://openid.net/specs/openid-connect-core-1_0.html#UserInfoError

```
5.3.3.  UserInfo Error Response
When an error condition occurs, the UserInfo Endpoint returns an Error Response as defined in Section 3 of OAuth 2.0 Bearer Token Usage [RFC6750]. (HTTP errors unrelated to RFC 6750 are returned to the User Agent using the appropriate HTTP status code.)

The following is a non-normative example of a UserInfo Error Response:

  HTTP/1.1 401 Unauthorized
  WWW-Authenticate: error="invalid_token",
    error_description="The Access Token expired"
```

Teleport should be updated to support both. Might be as simple as updating line 797 to the following.

```go
if trace.IsNotFound(err) || trace.IsAccessDenied(err) {
   [...]
}
```

### PR DESCRIPTION
**Pull request by atburke**
This PR allows `claimsFromUserInfo()` to gracefully handle errors from the UserInfo endpoint by returning just the ID token claims (same behavior as for an invalid endpoint).

Fixes #9768.

### CODE REVIEW

File: lib/auth/oidc.go [hunk 1]
```diff
@@ -174,7 +187,7 @@ func (a *Server) CreateOIDCAuthRequest(req services.OIDCAuthRequest) (*services.
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	oidcClient, err := a.getOrCreateOIDCClient(connector)
+	oidcClient, err := a.getOrCreateOIDCClient(connector, false)
```
Comment (codingllama) [inline lib/auth/oidc.go:190]:
```suggestion
	oidcClient, err := a.getOrCreateOIDCClient(connector, false /* insecure */)
```

File: lib/auth/oidc.go [hunk 2]
```diff
@@ -299,7 +312,7 @@ func (a *Server) validateOIDCAuthCallback(q url.Values) (*oidcAuthResponse, erro
 		return nil, trace.Wrap(err)
 	}
 
-	oidcClient, err := a.getOrCreateOIDCClient(connector)
+	oidcClient, err := a.getOrCreateOIDCClient(connector, false)
```
Comment (codingllama) [inline lib/auth/oidc.go:315]:
```suggestion
	oidcClient, err := a.getOrCreateOIDCClient(connector, false /* insecure */)
```

File: lib/auth/oidc_test.go [hunk 3]
```diff
@@ -86,48 +86,70 @@ func (s *OIDCSuite) TestCreateOIDCUser(c *check.C) {
...
 // trace.NotFound similar to an invalid userinfo endpoint. For these users,
 // all claim information is already within the token and additional claim
 // information does not need to be fetched.
-func (s *OIDCSuite) TestUserInfo(c *check.C) {
+func TestUserInfoBlockHTTP(t *testing.T) {
+	s := setUpSuite(t)
 	// Create configurable IdP to use in tests.
-	idp := newFakeIDP()
-	defer idp.Close()
+	idp := newFakeIDP(t, false)
```
Comment (codingllama) [inline lib/auth/oidc_test.go:108]:
```suggestion
	idp := newFakeIDP(t, false /* tls */)
```

So on for similar occurrences.

File: lib/auth/oidc.go [hunk 4]
```diff
@@ -125,6 +126,18 @@ func oidcConfig(conn types.OIDCConnector) oidc.ClientConfig {
 		// open id notifies provider that we are using OIDC scopes
 		Scope: apiutils.Deduplicate(append([]string{"openid", "email"}, conn.GetScope()...)),
 	}
+
+	if insecure {
+		conf.HTTPClient = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
```
Comment (codingllama) [inline lib/auth/oidc.go:134]: How is this change related to the issue? Is this just for testing?

If yes, could we change the tests so it's always TLS and revert this part?

File: lib/auth/oidc_test.go [hunk 5]
```diff
@@ -86,48 +86,70 @@ func (s *OIDCSuite) TestCreateOIDCUser(c *check.C) {
...
 	connector, err := types.NewOIDCConnector("test-connector", types.OIDCConnectorSpecV3{
 		IssuerURL:    idp.s.URL,
 		ClientID:     "00000000000000000000000000000000",
 		ClientSecret: "0000000000000000000000000000000000000000000000000000000000000000",
 	})
-	c.Assert(err, check.IsNil)
-	oidcClient, err := s.a.getOrCreateOIDCClient(connector)
-	c.Assert(err, check.IsNil)
+	require.NoError(t, err)
+	oidcClient, err := s.a.getOrCreateOIDCClient(connector, false)
```
Comment (codingllama) [inline lib/auth/oidc_test.go:117]:
```suggestion
	oidcClient, err := s.a.getOrCreateOIDCClient(connector, false /* insecure */)
```

So on for similar occurrences.

File: lib/auth/oidc.go [hunk 6]
```diff
@@ -649,6 +662,10 @@ func (a *Server) getClaims(oidcClient *oidc.Client, connector types.OIDCConnecto
 			log.Debugf("OIDC provider doesn't offer valid UserInfo endpoint. Returning token claims: %v.", idTokenClaims)
 			return idTokenClaims, nil
 		}
+		if trace.IsAccessDenied(err) {
```
Comment (codingllama) [inline lib/auth/oidc.go:665]: Should we specifically look for 401?

Unless I'm reading it wrong, It looks like AccessDenied includes anything that isn't a 2xx in this case ([lib/auth/oidc.go#595](https://github.com/gravitational/teleport/blob/2d958d69571f5c028ceb51e55ce9d01e9c520f2c/lib/auth/oidc.go#L595-L597)).

File: lib/auth/oidc.go (hunk 4 above)
Comment (atburke) [reply lib/auth/oidc.go:134]: Yes, this is for testing. However, `TestUserInfoBlockHTTP` specifically tests a non-TLS server.

File: lib/auth/oidc.go (hunk 6 above)
Comment (atburke) [reply lib/auth/oidc.go:665]: Right. According to [the spec](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoError) and [RFC 6750](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1), UserInfo can return any appropriate status code. I don't think we need to filter for 401 unless we want to take different actions for different codes.

File: lib/auth/oidc.go (hunk 4 above)
Comment (codingllama) [reply lib/auth/oidc.go:134]: Is that a feasible real-world scenario? I'm wary of adding code that has no production value, especially code that allows a relaxation of security parameters.

File: lib/auth/oidc.go (hunk 6 above)
Comment (codingllama) [reply lib/auth/oidc.go:665]: You have much better context here than me, so let me know if this doesn't make sense:

I think I'd rather check only for the handful of statuses that make sense/we know happen (400, 401, 403, or 405, according to the RFC), but still fail if we get something outside of that range, so if something too weird happens it won't be silent. If this comes and bites us again, then I would relax the check.

In any case, please add a comment in the code calling to attention the fact that are handling a various of statuses with this if; usually I would expect AccessDenied to be 401/403.

File: lib/auth/oidc.go (hunk 4 above)
Comment (atburke) [reply lib/auth/oidc.go]: You're right, there shouldn't be a non-testing scenario where we use this; I've moved this to tests only.

File: lib/auth/oidc.go (hunk 6 above)
Comment (atburke) [reply lib/auth/oidc.go:652]: That sounds fine. Is it worth adding custom errors in [trace](https://github.com/gravitational/trace) for 401 and 405, or should I stick with bundling them all under `AccessDenied`?

Comment (codingllama) [reply lib/auth/oidc.go:652]: No need to hold your PR on a trace change, I think. Either bundle in AccessDenied or return specific errors for different types, up to you - just mention in `claimsFromUserInfo`'s docs what AccessDenied means there, if you go that route.

File: lib/auth/oidc_test.go [hunk 7]
```diff
@@ -74,10 +75,40 @@ func (s *OIDCSuite) SetUpSuite(c *check.C) {
...
+	}
+	client, err := oidc.NewClient(conf)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaults.WebHeadersTimeout)
+	defer cancel()
+
+	go func() {
+		defer cancel()
+		client.SyncProviderConfig(connector.GetIssuerURL())
```
Comment (codingllama) [inline lib/auth/oidc_test.go:100]: Can we drop the goroutine and do this in the body of `createInsecureOIDCClient`?

File: lib/auth/oidc.go (hunk 4 above)
Comment (codingllama) [reply lib/auth/oidc.go]: Much better, thanks.

File: lib/auth/oidc_test.go [hunk 8]
```diff
@@ -74,10 +75,40 @@ func (s *OIDCSuite) SetUpSuite(c *check.C) {
...
+	}()
+
+	<-ctx.Done()
+	if ctx.Err() != nil {
```
Comment (gabrielcorado) [inline lib/auth/oidc_test.go:104]: I think this check is not necessary since the `Done` channel will be closed so that you can expect a non-nil error:

> [If Done is closed, Err returns a non-nil error explaining why:](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/context/context.go;l=102)

**Review by gabrielcorado — APPROVED** [review]

**Review by codingllama — APPROVED** [review]
